### PR TITLE
Support pytorch extensions

### PR DESCRIPTION
**Pull request by rdadolf**
PyTorch allows new operators to be registered dynamically in modules. Torch-mlir already makes it fairly straightforward to add support for new operators, and this commit just extends that support to allow new PyTorch ops to come from an external module.

Previously, there was no hook to allow extensions to be loaded before querying the op registry, so extension ops would never be generated. Moreover, a bug caused ops imported from an extension to be impossible to support via shape inference.

This does *not* allow ops to be dynamically loaded into torch-mlir. Torch-mlir must still be compiled with support built-in.

This should make tasks like [this one](https://github.com/llvm/torch-mlir/issues/462#issuecomment-989036106) in #462 (and to a lesser extent #504) somewhat easier.

### CODE REVIEW

**Comment by rdadolf** [comment]
@silvasean Since you originally wrote most of these pieces, I'd appreciated an opinion.

**Comment by powderluv** [comment]
drive-by thank you for the PR :D

**Comment by silvasean** [comment]
This looks awesome. Before committing this, I would like to somehow make it load-bearing for the project so it gets tested/etc. To do that, I think we should add a minimal custom op end-to-end, which I think would consist of:

1. build Python module defining custom torch op
2. add stuff in ods_gen and shape_lib_gen for that op
3. dtype refinement in RefineTypes
4. TorchToLinalg/CustomOpsExample.cpp

I think it could be as simple as a custom unary op -- something for folks to follow the bread crumbs when adding their own. I think the QPyTorch folks are going to want to follow this too.

**Comment by rdadolf** [comment]
I wrote and tested all of that (in fact, it's literally written for a unary no-op---seems we had the same idea), but two things stopped me from including it in the PR:

1. I had no idea where to put it for testing purposes.
2. It means the test op will be exposed to the "real" world, which seemed unclean at the time.

If you folks don't mind, I don't mind.

**Comment by silvasean** [comment]
> * I had no idea where to put it for testing purposes.

What form does the custom op take? Is it just a Python module? Or is there C++ code involved?

> * It means the test op will be exposed to the "real" world, which seemed unclean at the time.

Yeah, I'm sympathetic to this. However, PyTorch's op list is already pretty unclear and has all sorts of random stuff with varying levels of "core-ness" vs "custom-ness". I don't think that `torch_mlir_custom.nop` is that much more unclean than `aten._foo_cudnn_special_thing`

**Comment by rdadolf** [comment]
Sounds good. I'll start adding the rest of the code into the PR.

> What form does the custom op take? Is it just a Python module? Or is there C++ code involved?

My example was a python wrapper around a C++ extension, similar to the [tutorial from the PyTorch docs](https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html). It's small, but it does include a cmake file, requires compilation, etc.

The short version is:

Python side:
```python
import os
import torch

# Register custom.nop as a side-effect of importing this module.
current_dir = os.path.dirname(os.path.abspath(__file__))
lib = os.path.join(*[current_dir, os.pardir, os.pardir, 'custom_nop', 'libnop.so'])
torch.ops.load_library(lib)
```

C++ side:
```c++
#include <torch/script.h> // One-stop header.

#include <iostream>

torch::Tensor nop(torch::Tensor t) {
  std::cout << "no-op executed\n";
  return t;
}

TORCH_LIBRARY(custom, m) {
  m.def("nop(Tensor t) -> Tensor");
  m.impl("nop", &nop);
}
```

**Comment by silvasean** [comment]
Ah, ok. Then I think this can be a standalone build inside build_tools. Perhaps `build_tools/custom_op_example`. And for clarity, let's make sure that the "namespace" of the op is always `torch_mlir_custom_op_example`, e.g. `torch_mlir_custom_op_example.nop`.

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py
Comment (silvasean) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py:1059]: nit: format

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py
Comment (silvasean) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/shape_lib_gen.py:1060]: nit: `"` for strings.

File: lib/Conversion/TorchToLinalg/CustomOpExample.cpp
Comment (silvasean) [inline lib/Conversion/TorchToLinalg/CustomOpExample.cpp:40]: rewriter.replaceOp is more succinct.

File: python/torch_mlir/_torch_mlir_custom_op_example/torch_mlir_custom_op_example.cpp
Comment (silvasean) [inline python/torch_mlir/_torch_mlir_custom_op_example/torch_mlir_custom_op_example.cpp:6]: strictly speaking this is an `identity`, not a `nop` -- might make it a bit clearer.

File: build_tools/update_shape_lib.sh
Comment (silvasean) [inline build_tools/update_shape_lib.sh:2]: maybe add a "usage:" in the comment at the top of this file covering the new env vars.

File: lib/Conversion/TorchToLinalg/CustomOpExample.cpp
Comment (rdadolf) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp:40]: Doesn't `replaceOp` require an op argument to replace with? The intent here was to remove the identity op altogether, which means that uses of its return value must instead use the argument to the identity op.

File: lib/Conversion/TorchToLinalg/CustomOpExample.cpp
Comment (rdadolf) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp]: TIL about C++ reserved name formats.

We don't actually control the extension naming scheme. Obviously if someone's extending torch-mlir, they probably have some influence into the upstream extension package, but it still seems a bit wrong to have our code willingly generate illegal names under otherwise legal conditions.

I'll take an alternate view: what's the right thing to do if we *do* encounter a package with a leading underscore? Well, one solution is just to always generate legal code. I.e., when translating the Python names to C++, strip leading underscores. Technically, this collapses `_package_name` and `package_name` to `PackageName`, but... I think that's probably the right answer here, since the alternative is illegal C++ code.

The fix for that is a one-line change to `registry.py`. This has the added benefit of *always* preventing illegal names, which seems positive.

(I'll fix the string delimiters in `registry.py` that I overlooked in a bit. Currently debugging a crash that I can't reproduce locally.)

Comment (silvasean) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp]: What is it on the Torch side that requires the leading _ ? Is it the "namespace" of the extension? The .so file name?

To the degree that the leading _ is mandatory, then removing it should not result in collisions.

File: lib/Conversion/TorchToLinalg/CustomOpExample.cpp
Comment (silvasean) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp:40]: replaceOp takes a list of values (each corresponding to one result of the op being replaced). so you would pass in the operand value. You might be thinking of replaceOpWithNewOp

File: lib/Conversion/TorchToLinalg/CustomOpExample.cpp
Comment (rdadolf) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp]: > What is it on the Torch side that requires the leading _ ? Is it the "namespace" of the extension? The .so file name?

The extension package name, which ends up being concatenated with the op to give a unique op name. The .so file can be named whatever you wish, since it's entirely controlled by the extension and its build files.

Realistically, nothing about our name translation system is "mandatory", since it's all one-way. Torch-mlir choose how to change Python names into C++ names, and then those C++ names are used internally in torch-mlir. There's nothing that would prevent us from turning `aten.tanh` into `XXMLIRXXatenXXDOTXXtanhXXMLIRXX`, besides driving everyone batty reading it.

So if we think about the whole name translation system as a map, it might look like this:
```
Python Names -> C++ Names
-------------------------
package.op -> PackageOp
package_foo.op_foo -> PackageFooOpFoo
foo.bar.op -> FooBarOp
```

The case we're talking about is this:
```
_package.op -> PackageOp (proposed, currently illegal as _PackageOp)
package.op -> PackageOp
foo._bar.op -> Foo_BarOp
```

Both `_package.op` and `package.op` are technically legal Python names which can register op extensions, so under the current system, they can generate illegal C++ identifiers (without warning). I suppose we could alter the renaming scheme to change leading underscores into something else like `_package.op -> UnderscorePackageOp`, but this seems to violate the principle of least surprise.

At the end of the day, the problem is that this part of the legal Python namespace is a superset of the legal C++ names, so there's always going to be a bit of a rub. We kind of have to choose one of the three:

1. Allow for the possibility of illegal C++ names.
2. Allow for the possibility of namespace collisions.
3. Mangle the python names

My opinion was that 2. is the least bad/intrusive, but I'm happy to reconsider.

Comment (silvasean) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp]: These two quotes seem to be saying opposite things:

1.
> > What is it on the Torch side that requires the leading _ ? Is it the "namespace" of the extension? The .so file name?
> The extension package name, which ends up being concatenated with the op to give a unique op name. 

2.
> Both _package.op and package.op are technically legal Python names which can register op extensions


Is the leading `_` mandatory?

- If it is mandatory, then dropping the leading _ should never result in a collision
- If it is not mandatory, then why don't we just name our package without it?

Either way, I think that your option 2 makes sense.

Comment (rdadolf) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp]: Sorry, I'll try to be a more clear:

PyTorch extensions aren't necessarily written for torch-mlir, so I'm assuming we don't control the package naming scheme in the general case. The leading `_` is not mandatory for *our* example package and we could definitely remove it, but there's nothing that prevents anyone from developing a PyTorch extension that is named that way. If they do, we don't have much choice in what is fed to us. The name will be whatever the name is. So in that sense, we only have a choice about what to do with what's coming down the pipe.

Honestly, I don't have any kind of census about how often leading-underscore names appear in PyTorch op extensions (I suspect it is not common), so this could really all be unimportant either way.

Comment (rdadolf) [reply lib/Conversion/TorchToLinalg/CustomOpExample.cpp]: Sounds good. Can always come back with another fix if it becomes a problem.

**Comment by rdadolf** [comment]
All set for merge, I think. Happy to take any last comments or requests.

**Review by silvasean — APPROVED** [review]
Awesome, thanks Bob!!

Can you squash the history before merging to main?

**Comment by rdadolf** [comment]
> Can you squash the history before merging to main?

Done. I don't have write access to the repo, so I'll need someone to hit the button. Thanks!

**Comment by silvasean** [comment]
> > Can you squash the history before merging to main?
> 
> Done. I don't have write access to the repo, so I'll need someone to hit the button. Thanks!

Just invited you. Can't believe you didn't have it!

**Pull request merged**
